tableaudocumentapi/workbook.py

```diff
@@ -95,12 +109,45 @@ def save_as(self, new_filename):
     # Private API.
     #
     ###########################################################################
-    def _prepare_datasources(self, xmlRoot):
+    @staticmethod
+    def _prepare_datasource_index(datasources):
+        retval = weakref.WeakValueDictionary()
+        for datasource in datasources:
+            retval[datasource.name] = datasource
+
+        return retval
+
+    @staticmethod
+    def _prepare_datasources(xml_root):
         datasources = []
 
         # loop through our datasources and append
-        for datasource in xmlRoot.find('datasources'):
+        for datasource in xml_root.find('datasources'):
             ds = Datasource(datasource)
             datasources.append(ds)
 
         return datasources
+
+    @staticmethod
+    def _prepare_worksheets(xml_root, ds_index):
+        worksheets = []
+        worksheets_element = xml_root.find('.//worksheets')
```

Comment:
okay I wrote a test for "an empty workbook should not throw errors" to catch all of the places where find returns None but we are assuming that it can't be None, and removed the extraneous check for None in the case of dependencies.

the difference is find returns None when nothing is found, but findall returns an empty list, which we don't need to check since iterating over an empty list is a nop anyway.